src/interface.jl

```
@@ -59,45 +63,90 @@ Return a vector of length `D` of `BoundaryCondition` objects, one for each direc
 """
 function boundary_conditions end
 
+"""
+    species_type(::AbstractSystem)
+
+Return the type used to represent a species or atom.
+"""
+function species_type end
+
 periodicity(sys::AbstractSystem) = [isa(bc, Periodic) for bc in boundary_conditions(sys)]
 
-# Note: Can't use ndims, because that is ndims(sys) == 1 (because of indexing interface)
+"""
+    n_dimensions(::AbstractSystem)
+    n_dimensions(atom)
+
+Return number of dimensions.
+"""
 n_dimensions(::AbstractSystem{D}) where {D} = D
+# Note: Can't use ndims, because that is ndims(sys) == 1 (because of indexing interface)
 
 # indexing and iteration interface...need to implement getindex and length, here are default dispatches for others
 Base.size(s::AbstractSystem) = (length(s),)
 Base.setindex!(::AbstractSystem, ::Int) = error("AbstractSystem objects are not mutable.")
```

Does this need to throw a generic error? If not undefined (or perhaps mentioned in a comment) it would throw a `MethodError`. The error message as written may not be correct anyway, since someone could make a `mutable struct` system.

---

Agreed. I removed the message completely (that's how StaticArrays does it for example).